### ISSUE TITLE
Can't send Gifs, error code 5 

### ISSUE THREAD

**Issue by Hwmpunk** [FossifyOrg/Messages#45]
### Checklist

- [X] I can reproduce the bug with the latest version given [here](https://github.com/FossifyOrg/Messages/releases/latest).
- [X] I made sure that there are **no existing issues** - [open](https://github.com/FossifyOrg/Messages/issues) or [closed](https://github.com/FossifyOrg/Messages/issues?q=is%3Aissue+is%3Aclosed) - to which I could contribute my information.
- [X] I made sure that there are **no existing discussions** - [open](https://github.com/FossifyOrg/Messages/discussions) or [closed](https://github.com/FossifyOrg/Messages/discussions?discussions_q=is%3Aclosed) - to which I could contribute my information.
- [X] I have read the FAQs inside the app (Menu -> About -> FAQs) and my problem isn't listed.
- [X] **I have taken the time to fill in all the required details. I understand that the bug report will be dismissed otherwise.**
- [X] This issue contains only one bug.
- [X] I have read and understood the [contribution guidelines](https://github.com/FossifyOrg/Messages/CONTRIBUTING.md).

### Affected app version

1.0.0

### Affected Android/Custom ROM version

Graphene OS

### Affected device model

Pixel 8 pro

### How did you install the app?

None

### Steps to reproduce the bug

Sending gif text results in cannot send, touch to retry. Then error code 5 pops up. 

### Expected behavior

Can't send Gifs 

### Actual behavior

Error code 5

**Comment by Aga-C** [comment]
Are you able to send normal photos?

**Comment by naveensingh** [comment]
Error code 5 is usually thrown when you are sending something that exceeds carrier limits. It may be that you have sent too many MMS messages or you are sending a file that is too big. The error messages need to be more user-friendly. 

@Hwmpunk Can you please try sending a very small file no more than 20-100kbs?

**Comment by Hwmpunk** [comment]
> Error code 5 is usually thrown when you are sending something that exceeds carrier limits. It may be that you have sent too many MMS messages or you are sending a file that is too big. The error messages need to be more user-friendly.
> 
> @Hwmpunk Can you please try sending a very small file no more than 20-100kbs?

I sent a 191kb gif and it worked. The one that didn't was exactly 1mb. The stock Graphene messaging app can send larger Gifs and videos under 2mb. For some reason your app has a smaller limit of less than 1mb. But yes, there's a limit on standard mms.

**Comment by naveensingh** [comment]
The app always tries to resize normal images to the MMS limit set in app settings. It doesn't do that for GIFs (yet) so that explains why your stock messages app can send it by compressing it. Have you tried sending a 2MB JPEG or PNG image? It shouldn't fail.

**Comment by Aga-C** [comment]
I think, while we don't have such feature, we could temporarily add a warning, that file is too large instead of showing cryptic "error code 5".

**Comment by naveensingh** [comment]
Yes, some of the error codes have some useful descriptions available but I didn't update error code 5 for some reason that I can't remember. Currently, 3 errors have user-friendly messages out of all the error codes available at [SmsManager.java:2399](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/telephony/java/android/telephony/SmsManager.java;l=2399?q=android.telephony.SmsManager&ss=android%2Fplatform%2Fsuperproject%2Fmain)

I guess we can close this as the original issue is not an 'issue'. We'll later add support for compressing different formats like GIFs.

**Comment by Hwmpunk** [comment]
> The app always tries to resize normal images to the MMS limit set in app settings. It doesn't do that for GIFs (yet) so that explains why your stock messages app can send it by compressing it. Can you try sending a ~2MB JPEG or PNG image? It shouldn't fail.

I tried sending a 2.41mb photo and still received the error code 5 so it appears this app does not automatically reduce the size on any type of media.

**Comment by naveensingh** [comment]
What is the value of the "Resize sent MMS images" setting? Is it 300KB? Can you try lowering it down to 200KB and then sending the message?

**Comment by naveensingh** [comment]
In fact, you can send images of up to 100MBs and they will be compressed under the set MMS limit: https://github.com/SimpleMobileTools/Simple-SMS-Messenger/pull/589

**Comment by DarkSirrush** [comment]
Just to add to this, I can't send images with carrier limit disabled with this app, however I can send images with limit disabled with Textra. This is on a stock Samsung flip 4, v1.0.0 of app installed via FDroid Izzy repo.

It looks my carrier does compress the image themselves however, as the received image goes from 2.96mb to 294kb. Maybe that is related?

Also note, when setting carrier limit to 300kb, it allows me to send, however the file is compressed to 158kb and looks significantly worse than the one sent via Textra with "no limit". See screenshot below:
![Screenshot_20240124_122437_Messages_20240124122858](https://github.com/FossifyOrg/Messages/assets/79956291/e0a3d788-0a13-4276-856c-32610a8f8b58)

**Comment by Hwmpunk** [comment]
> What is the value of the "Resize sent MMS images" setting? Is it 300KB? Can you try lowering it down to 200KB and then sending the message?

Confirmed that setting the limit to 2mb allowed a 6mb image to be sent. However, the image received is now 766kb

> What is the value of the "Resize sent MMS images" setting? Is it 300KB? Can you try lowering it down to 200KB and then sending the message?

Confirmed that setting image limit to 2mb allowed a 6mb image to be sent. However, the image was reduced to 766kb. Would be nice if it could retain as much quality as possible.

**Comment by naveensingh** [comment]
> I can't send images with carrier limit disabled with this app, however I can send images with limit disabled with Textra.
> It looks my carrier does compress the image themselves however, as the received image goes from 2.96mb to 294kb.

An alternative explanation is that Textra is always compressing images to under 300KB or by some factor and is ignoring the set MMS limit.

> Confirmed that setting the limit to 2mb allowed a 6mb image to be sent. However, the image received is now 766kb

Great! This means there is no MMS bug and it's the image compressor that needs to be improved.

**Comment by DarkSirrush** [comment]
> An alternative explanation is that Textra is always compressing images to under 300KB or by some factor and is ignoring the set MMS limit.
> 

I sure hope not, and when I sent the same image with a 2mb limit in textra I got a 293.8kb file. I might test receiving images with a few sms apps later to confirm it's my carrier later.

**Comment by naveensingh** [comment]
Addressing this properly requires:
 - [ ] Support for GIF compression
 - [ ] Support for video compression
 - [ ] Improved (more accurate) image compression.

**Comment by breversa** [comment]
Error code 5 also happens when your APN doesn’t have a proper MMS configuration (missing MMSC, MMS proxy, MMS port, and/or missing "mms" in "APN type").

**Comment by asimons04** [comment]
> Error code 5 also happens when your APN doesn’t have a proper MMS configuration (missing MMSC, MMS proxy, MMS port, and/or missing "mms" in "APN type").

Thanks for this.  That was the problem in my case.

My device is on Mint Mobile (T-Mobile MVNO).  MMS used to work fine, but at some point, it auto-provisioned with T-Mobile APN settings, and I'm guessing that's when MMS in Fossify broke.

Everything (VoLTE, data, SMS, receiving MMS) worked otherwise, but Fossify Messenger would not send any media over MMS regardless of the size (the stock Messages app would, though).  MMS messages were able to be received, though.

When I checked just now, the T-Mobile APNs were split:  One for data and a separate one for MMS.  The stock Messages app seemed to be able to cope with this, but Fossify could not.

Not sure if I solved or worked around the problem, but I manually created the Mint APN based on the settings they provide on their support page (which has a single APN for all services).  Fossify Messenger now works correctly when sending media over MMS.

Hope this helps someone.

**Comment by Valeryan24** [comment]
Hi, first thanks a lot for your great work on Fossify apps (Messages, Gallery) - it's much appreciated !

I also have this bug since months (years), and I thought the recent posts about APN configuration would solve it, but unfortunately no…

I have a Samsung Galaxy S22+ and french mobile operator Free : I tried many configurations (manual with many different options, reset to default), nothing works.

Base APN type is well : "default,mms"

https://assistance.free.fr/univers/mobile-free/mon-mobile/premiers-pas/device/samsung/galaxy-s22-plus/topic/mms/configurer-mms-manuellement/9

I still can't send MMS via Fossify Messages, I always get "Error 5". I can receive MMS with Fossify, and I can send MMS with Google Messages, so the APN configuration works, but not for Fossify !

I don't understand where the problem is - but if it can help, I can make some tests.

**Comment by breversa** [comment]
> I have a Samsung Galaxy S22+ and french mobile operator Free : I tried many configurations (manual with many different options, reset to default), nothing works.
> 
> Base APN type is well : "default,mms"
> 
> https://assistance.free.fr/univers/mobile-free/mon-mobile/premiers-pas/device/samsung/galaxy-s22-plus/topic/mms/configurer-mms-manuellement/9
> 
> I still can't send MMS via Fossify Messages, I always get "Error 5". I can receive MMS with Fossify, and I can send MMS with Google Messages, so the APN configuration works, but not for Fossify !

With Free Mobile, you need to set up two APNs:
- one using "free" as APN, and "default,supl" as APN type ("supl" may be irrelevant for that issue)
- one using "freemms" as APN, and "mms" as APN type

**Comment by Valeryan24** [comment]
Hi, I well have the 2 APN - the second "Free MMS" can't be selected but is well listed.

I tried many manual configurations, included the one as you and Free suggest : first APN Free with "default, supl, hipri" and second "Free MMS", type only mms

https://assistance.free.fr/articles/1548

Then reset to the default configuration : Samsung put the 2 APN, Free with type "default,mms" and Free MMS with type only mms.

I also tried to add "Authentification type" PAP to the MMS APN as Free suggests.

Of course I restarted the smartphone after each modification.

No config works, always the "Error 5" when trying to send a MMS.

**Comment by breversa** [comment]
Huh… I’ve had my fair share of APN woes with Free Mobile, but I can at least tell you this:
- having two APNs with `type=mms` makes no sense
- having two APNs but only one selectable makes sense (it’s the one with `type=default`)
- your default Samsung configuration makes no sense. Which ROM are you using? Stock?

**Comment by Valeryan24** [comment]
Yes, stock ROM, Android 15. And if I switch to Google message with the same "new default "configuration, MMS sending works fine and is well received by the recipient…

Before I tried many manual configs, I already had the 2 APN, but if I remember well, the first one (Free) had only type "default" and nothing in MMSC, but it still didn't work with Fossify (it worked with Samsung and Google messages apps). Now, after having "reset to default", yes it added "mms" also to the first APN "Free"…

**Comment by breversa** [comment]
Last suggestion for tonight: once you’ve reset the configuration to default, remove `type=sms` from the the first APN "Free".

Also, set back the MMS size to the default (or minimum) value in Fossify Messages.

Then reboot and try again.

**Comment by Valeryan24** [comment]
OMG thanks a lot, it works !

Just, if I select for the "Resize MMS image" option "no limit" or "2 MB" in Fossify, I still get error 5…

But if I select 100 Kb or 1 MB, MMS is finally well sent this time 👍

**Comment by breversa** [comment]
> OMG thanks a lot, it works !
> 
> Just, if I select for the "Resize MMS image" option "no limit" or "2 MB" in Fossify, I still get error 5…
> 
> But if I select 100 Kb or 1 MB, MMS is finally well sent this time 👍

Glad I could help!

It appears that every phone carrier has an undocumented limit to the MMS message size. Free Mobile’s seem to be between 1 MB and 2 MB.

Maybe we should try and document them somewhere?

**Comment by naveensingh** [comment]
There's a plan to map these error codes to user-friendly messages: https://github.com/FossifyOrg/Messages/issues/315

> Maybe we should try and document them somewhere?

Ideally, the app should handle limits in-app. The [mms config files are outdated](https://github.com/FossifyOrg/Messages/issues/310#issuecomment-2630730047), but could be parsed at runtime to automatically apply the carrier-specific limits instead of using a preference. There's also the [CarrierConfigManager](https://developer.android.com/reference/android/telephony/CarrierConfigManager) API (or [SmsManager.getCarrierConfigValues()](https://developer.android.com/reference/android/telephony/SmsManager#getCarrierConfigValues())), but I haven't checked if it provides up-to-date information.

**Comment by AntainRhys** [comment]
For me, all I had to do was change *Resize sent MMS messages* from *No limit* to *2MB*.
